src/detect-ftp-mode.c

```diff
@@ -0,0 +1,186 @@
+    return 1 ? ftpmoded->active == tx->active : 0;
+}
+
+/**
+ * \brief this function will free memory associated with DetectFtpModeData
+ *
+ * \param ptr pointer to DetectFtpModeData
+ */
+static void DetectFtpModeFree(DetectEngineCtx *de_ctx, void *ptr)
+{
+    SCFree(ptr);
+}
+
+/**
+ * \brief This function is used to parse ftp.mode options passed via ftp.mode keyword
+ *
+ * \param str Pointer to the user provided ftp.mode options
+ *
+ * \retval  pointer to DetectFtpModeData on success
+ * \retval NULL on failure
+ */
+static DetectFtpModeData *DetectFtpModeParse(const char *optstr)
+{
+    DetectFtpModeData *ftpmoded = NULL;
+    char arg1[8] = "";
+    size_t pcre2len;
+    pcre2_match_data *match = NULL;
+
+    int ret = DetectParsePcreExec(&parse_regex, &match, optstr, 0, 0);
+    if (ret != 2) {
+        SCLogError("parse error, ret %" PRId32 "", ret);
+        goto error;
+    }
+
+    pcre2len = sizeof(arg1);
+    int res = pcre2_substring_copy_bynumber(match, 1, (PCRE2_UCHAR8 *)arg1, &pcre2len);
+    if (res < 0) {
+        SCLogError("pcre2_substring_copy_bynumber failed");
+        goto error;
+    }
+
+    ftpmoded = SCCalloc(1, sizeof(DetectFtpModeData));
+    if (unlikely(ftpmoded == NULL))
+        goto error;
+    if (!strcmp(arg1, "active")) {
+        ftpmoded->active = true;
+    } else if (!strcmp(arg1, "passive")) {
+        ftpmoded->active = false;
+    } else {
+        SCLogError("Invalid command value");
+        goto error;
+    }
+
+    pcre2_match_data_free(match);
+    return ftpmoded;
+
+error:
+    if (match) {
+        pcre2_match_data_free(match);
+    }
+    if (ftpmoded)
+        SCFree(ftpmoded);
+    return NULL;
+}
+
+static int DetectFtpModeSetup(DetectEngineCtx *de_ctx, Signature *s, const char *str)
+{
+    if (DetectSignatureSetAppProto(s, ALPROTO_FTP) != 0)
+        return -1;
+
+    DetectFtpModeData *ftpmoded = DetectFtpModeParse(str);
+    if (ftpmoded == NULL)
+        return -1;
+
+    if (SigMatchAppendSMToList(de_ctx, s, DETECT_FTP_MODE, (SigMatchCtx *)ftpmoded,
+                g_ftp_mode_buffer_id) == NULL) {
+        DetectFtpModeFree(de_ctx, ftpmoded);
+        return -1;
+    }
+
+    return 0;
+}
+
+void DetectFtpModeRegister(void)
+{
+    /* ftp.mode sticky buffer */
+    sigmatch_table[DETECT_FTP_MODE].name = KEYWORD_NAME;
+    sigmatch_table[DETECT_FTP_MODE].desc = "sticky buffer to match on the FTP mode buffer";
+    sigmatch_table[DETECT_FTP_MODE].url = "/rules/" KEYWORD_DOC;
+    sigmatch_table[DETECT_FTP_MODE].Setup = DetectFtpModeSetup;
+    sigmatch_table[DETECT_FTP_MODE].AppLayerTxMatch = DetectFtpModeMatch;
+    sigmatch_table[DETECT_FTP_MODE].Free = DetectFtpModeFree;
+
+    DetectAppLayerInspectEngineRegister(
+            BUFFER_NAME, ALPROTO_FTP, SIG_FLAG_TOCLIENT, 0, DetectEngineInspectGenericList, NULL);
+
+    DetectAppLayerInspectEngineRegister(
+            BUFFER_NAME, ALPROTO_FTP, SIG_FLAG_TOSERVER, 0, DetectEngineInspectGenericList, NULL);
```

Comment:
I think this will need updating a bit later, as the tx progress handling in ftp doesn't appear to be great. We need:
- per direction tx state tracking
- deduplication of "buffer" id's here.

E.g. ftpbounce registers a generic `ftp_request` buffer with the same properties. I think there is now a `ftp:request_started:generic` as well?

Reply:
Coming up with a patch for the tx progress issue.

Reply:
https://github.com/OISF/suricata/pull/13143